idle-crafting.lua

```diff
@@ -0,0 +1,495 @@
+    print('idle crafting: running main loop')
+    checkForWorkshop()
+    if not enabled then
+        return
+    end
+    local num_watched = {}
+    local watching = false
+
+    ---@type table<integer,boolean>[]
+    watched = {}
+    for idx, _ in ipairs(thresholds) do
+        watched[idx] = {}
+        num_watched[idx] = 0
+    end
+
+    for _, unit in ipairs(dfhack.units.getCitizens(true, false)) do
+        for idx, threshold in ipairs(thresholds) do
+            if getCraftingNeed(unit) > threshold then
+                watched[idx][unit.id] = true
+                num_watched[idx] = num_watched[idx] + 1
+                watching = true
+                goto continue
+            end
+        end
+        ::continue::
+    end
+    print(('watching %s dwarfs with crafting needs'):format(
+        table.concat(num_watched, '/')
+    ))
+
+    if watching then
+        repeatutil.scheduleUnlessAlreadyScheduled(GLOBAL_KEY .. 'unit', 53, 'ticks', unit_loop)
+    end
+end
+
+---enable main loop
+---@param enable boolean|nil
+local function start(enable)
+    enabled = enable or enabled
+    if enabled then
+        repeatutil.scheduleUnlessAlreadyScheduled(GLOBAL_KEY .. 'main', 8419, 'ticks', main_loop)
+    end
+end
+
+--- Handles automatic loading
+dfhack.onStateChange[GLOBAL_KEY] = function(sc)
+    if sc == SC_MAP_UNLOADED then
+        enabled = false
+        return
+    end
+
+    if sc ~= SC_MAP_LOADED or df.global.gamemode ~= df.game_mode.DWARF then
+        return
+    end
+
+    load_state()
+    start()
+end
+
+--
+-- Overlay to select workshops
+--
+
+
+IdleCraftingOverlay = defclass(IdleCraftingOverlay, overlay.OverlayWidget)
+IdleCraftingOverlay.ATTRS {
+    desc = 'Adds a UI to the Workers tab too enable idle crafting.',
+    default_pos = { x = -42, y = 41 },
+    default_enabled = true,
+    viewscreens = {
+        'dwarfmode/ViewSheets/BUILDING/Workshop/Craftsdwarfs/Tasks',
+    },
+    frame = { w = 55, h = 1 },
+}
+
+function IdleCraftingOverlay:init()
+    self:addviews {
+        widgets.CycleHotkeyLabel {
+            view_id = 'leisure_toggle',
+            frame = { l = 0, t = 0 },
+            label = 'Allow idle dwarfs to satisfy crafting needs:',
+            key = 'CUSTOM_L',
+            options = {
+                { label = 'yes', value = true, pen = COLOR_GREEN },
+                { label = 'no',  value = false },
+            },
+            initial_option = 'no',
+            on_change = self:callback('onClick'),
+        }
+    }
+end
+
+function IdleCraftingOverlay:onClick(new, _)
+    local workshop = dfhack.gui.getSelectedBuilding(true)
+    allowed[workshop.id] = new or nil
+    if new and not enabled then
+        start(true)
+    end
+    if not next(allowed) then
+        stop()
+    end
+    persist_state()
+end
+
+function IdleCraftingOverlay:onRenderBody(painter)
+    local workshop = dfhack.gui.getSelectedBuilding(true)
+    if not workshop then
+        return
+    end
+    persist_state()
```

Comment:
is this persist_state necessary?

Reply:
If anything, this should be a `load_state()`. Unless I messed up somewhere else, it's probably fine to just remove it and rely on `allowed` to be up to date.